EIPS/eip-5843.md

```diff
@@ -0,0 +1,448 @@
+
+`TOMONTX` takes two stack items as input `(top of stack) input, output`.  The most significant byte of each input is an EVMMAX slot (`input_slot`/`output_slot`).  If `evmmax_state == None` or the offset of the highest byte of the ranges referenced by the inputs is outside the bounds of allocated memory, consume all call gas and return to the caller in an exceptional state. Interpret the memory range associated with `input_slot` as little-endian value `input`, compute `mulmont(input, evmmax_state.r_squared, mod)` and place the result in the memory range referenced by `output_slot`.
+
+Conversion from Montgomery form `x_mont` to normal form can be computed with `mulmont(x_mont, 1, mod)`.
+
+## Rationale
+
+The choice of using an offset-based approach where operations happen on values in memory overcomes the need for stack manipulation and enables a wider variety of bit-widths than currently economical in the EVM.
+
+The choice of 64-bit aligned little-endian values is made to match the hardware of most machines removing the need for byte-swapping.
+
+### Benchmarks
+
+Benchmarks were performed on worst-case inputs.  Error bars represent one standard deviation with a sample size of 10 runs of a `go -bench=...` invocation.
+
+Fitting gas models is WIP.  I assume `MULMONTMX` will need a model which fits lower limbs closely while having subquadratic asymptotic behavior.
+![](https://github.com/jwasinger/evmmax-bench-plot/raw/master/charts/mulmontx_low.png)
+For reference a benchmark from https://github.com/holiman/uint256 of `MULMOD` doing 256-bit modular multiplication with inputs that are already reduced:
+```
+BenchmarkMulMod/mod256/uint256-4        	 6665253	       177.5 ns/op
+```
+![](../assets/eip-5843/mulmontx_med.png)
+![](../assets/eip-5843/mulmontx_all.png)
+
+`ADDMODMX` and `SUBMODMX`:
+![](../assets/eip-5843/addmodx_low.png)
+![](../assets/eip-5843/addmodx_med.png)
+![](../assets/eip-5843/addmodx_all.png)
+
+![](../assets/eip-5843/submodx_low.png)
+![](../assets/eip-5843/submodx_med.png)
+![](../assets/eip-5843/submodx_all.png)
+
+`SETMODMX` should have linear-ish complexity.  Modular inverse appears to be `O(logN)` and my current implementation allocates a lot of memory.
+![](../assets/eip-5843/submodx_low.png)
+![](../assets/eip-5843/submodx_med.png)
+![](../assets/eip-5843/submodx_all.png)
+
+`SETMODX` computes Montgomery constants used by generic `mulmont` and algorithms optimized for smaller limb counts.  As a result `SETMODX` is more expensive than it could be, especially for low limbs.  This choice is made to avoid explicitly including a pivot point in the protocol, as it may vary slightly depending on the implementation.
+
+### Other TODOs to note in the rationale:
+* choice of little-endian limbs made to match the hardware of most (all?) Ethereum nodes, removing the need for endianness-swapping individual limbs.
```

Comment:
It is unclear what limbs are in the specs.

In arbitrary precision bigints libraries, the usual _internal_ representation is:
- metadata about number of limbs, and whether the number is positive or negative
- limbs, from the least significant to the most significant words (eg. limb-endianness is little-endian)
- words, the endianness is hardware endianness (eg. word-endianness is little-endian)
- the size of the word (32 or 64 bits) is also hardware dependent.

That is common across GMP, Java, Go, LLVM internal bigint lib (in APint.h).
However that internal representation is typically inaccessible by third-parties, and you cannot cast to it to achieve the zero-compute that is aimed there.

The other usual representation is called "octet string" is just big-endian serialization fo the integer, usually called I2OSP (Integer to Octet String) from https://www.rfc-editor.org/rfc/rfc3447#section-4.1
All library supports this with an _external API_.

Furthermore, all CPUs and compilers offer __builtin_bswap64 for endianness swapping. And in any case, endianness swapping is virtually costless:
1. On Haswell CPUs (2013) or later, the BSWAP 64-bit instruction takes 2 cycles (less than 1 nanosecond on a 3GHz CPU) and can be issued twice per cycle (https://www.agner.org/optimize/instruction_tables.pdf)
2. Retrieving data from L2 cache takes 7ns or from RAM takes 100ns (https://gist.github.com/jboner/2841832)
3. data load and swapping have no dependencies so while swapping is taking place, the next limb can be retrieved.
4. so due to endian-swapping "latency hiding", converting endianness has the same cost as copying data + ~1ns

So in conclusion, little-endian limbs require the bigints library to expose some way to cast/copy directly into their internals, which is unlikely due to metadata, input validation, etc AND the difference with using standard octet-string is not even a nanosecond on 10-year old hardware.

Hence I don't think it's worthwhile while octet-string is supported in all bigint tools.

Reply:
The EVMMAX value format is 64bit-aligned, little endian.

I can see how the references to _limbs_ are confusing.  I'm going to revise the document to delineate between `evmmax_val_size` (the size of EVMMAX values in 64bit increments) and _limbs_ (a base-`1<<system_word_size_bits` value):  _limbs_ should only be referred to in the context of the example pseudo-code.

Good to hear that byteswapping overhead should be minimal.  If this is the case, then I think the format should be modified to be 64bit-aligned big-endian.  This also has the benefit of making the EVMMAX opcodes more interoperable with other EVM ops.